### ISSUE TITLE
Sort out completing-read

### ISSUE THREAD

**Issue by gracjan** [haskell/haskell-mode#833]
Completing read is a must-have these day. `haskell-mode` has some mechanisms to provide it, those need a clean up. 

```el
grep complet.*read *.el
haskell-cabal.el:;;   (let ((section (completing-read
haskell-complete-module.el:(defun haskell-complete-module-read (prompt candidates)
haskell-customize.el:(defcustom haskell-completing-read-function 'ido-completing-read
haskell-customize.el:          (function-item :tag "ido" :value ido-completing-read)
haskell-customize.el:          (function-item :tag "helm" :value helm--completing-read-default)
haskell-customize.el:          (function-item :tag "completing-read" :value completing-read)
haskell-load.el:                          (haskell-complete-module-read "Module: " modules)))
haskell-modules.el:                 (haskell-complete-module-read
haskell-session.el:    (let* ((session-name (funcall haskell-completing-read-function
haskell.el:           (let ((module (haskell-complete-module-read
haskell.el:     (funcall haskell-completing-read-function "Cabal command: "
```

These is some special code `haskell-complete-module-read`, what does it do?

To do:
- [ ] use `funcall haskell-completing-read-function` everywhere
- [ ] remove `haskell-complete-module-read` (Note: need to check what this really is)
- [ ] find out what is going on in `haskell-complete-module.el`. Either write proper documentation and test cases for it or remove the thing.

**Comment by bergey** [comment]
I think that we should change `haskell-add-import` to not use any of the other code in `haskell-complete-module` or `haskell-modules`, remove the import-completion behavior from `haskell-mode-contextual-space`, and deprecate the remaining code.  If this seems reasonable, I'll make a PR.

It seems that `haskell-complete-module` has a completing-read function, which does .-separated prefix completion.  `haskell-modules` finds candidate module names by grepping the pkg-store.  It's not as good as `haskell-completions-sync-completions-at-point`.

**Comment by gracjan** [comment]
Seems reasonable.

Please add tests for every functionality that you implement.

**Comment by saevarb** [comment]
I decided that I was tired of the completion functionality(particularly when completing modules) and decided to play around. I replaced the module completion functionality with code that checks if `(require 'helm nil 'noerror)` fails and uses helm if it doesn't. It seems to work just fine. However, I'm assuming from this issue that the problem should probably handled in a more robust way.

If I wanted to implement support for helm(and possibly other completion libraries?), where do I start? Is that something you want? The current functionality is pretty poor, and seeing as many users already have helm or ido or some such installed, it would probably be a good idea to add support for it when it is available.

**Comment by gracjan** [comment]
Yes, we would like this. Can you make a rought PR so we can discuss the code?

Note that it must give a way for people that do not use helm to use whatever they want.
